### PR TITLE
Use golangci-lint

### PR DESCRIPTION
**Pull request by denis-tingajkin**
Signed-off-by: denis-tingajkin <denis.tingajkin@xored.com>

## Description
Fix for https://github.com/networkservicemesh/networkservicemesh/issues/1044

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [ ] Added integration testing to cover
- [ ] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -23,10 +23,11 @@ jobs:
             ./scripts/go-mod-download.sh
       - run: "! go fmt ./... 2>&1 | read"
       - run:
-          name: golint
+          name: lint
           command: |
-            go get golang.org/x/lint/golint
-            golint ./...
+            go get github.com/golangci/golangci-lint/cmd/golangci-lint
```
Comment (nickolaev) [inline .circleci/config.yml]: Hmm, shall we use a pinned version of the tool, to have a better reproducibility and avoid occasional breakage?

File: controlplane/pkg/nseregistry/nse_registry.go [hunk 2]
```diff
@@ -38,8 +38,8 @@ func store(values ...string) string {
 			res += "\t"
 		}
 		ss := strings.Replace(s, "\t", "\\t,", -1)
-		ss = strings.Replace(s, "\n", "\\n,", -1) // Just in case
-		ss = strings.Replace(s, "\r", "\\r,", -1) // Just in case
+		ss = strings.Replace(ss, "\n", "\\n,", -1) // Just in case
+		ss = strings.Replace(ss, "\r", "\\r,", -1) // Just in case
```
Comment (nickolaev) [inline controlplane/pkg/nseregistry/nse_registry.go]: That is a really good catch.

File: .circleci/config.yml [hunk 3]
```diff
@@ -26,8 +26,7 @@ jobs:
           name: lint
           command: |
             go get github.com/golangci/golangci-lint/cmd/golangci-lint
-            echo ${CIRCLE_BRANCH}
-            golangci-lint run --new-from-rev=${CIRCLE_BRANCH} ./...
+            golangci-lint run --new-from-rev=origin/master
```
Comment (denis-tingajkin) [inline .circleci/config.yml]: Here I've added possible to ignore most existing problems. 
I expect `golangci-ling` will be running only with diff

Comment (denis-tingajkin) [reply .circleci/config.yml]: Also, I'll have provided a commit with fixing the most critical issues found by linters

Comment (nickolaev) [reply .circleci/config.yml]: Yep, we can create an issue to fix the rest and label it a `good new starter issue` or such.

File: docs/guide-build.md [hunk 4]
```diff
@@ -206,3 +206,24 @@ If you want to run it locally, you need to [install shellcheck](https://github.c
 ## Canonical source on how to build
 
 The [.circleci/config.yml](https://github.com/networkservicemesh/networkservicemesh/blob/master/.circleci/config.yml) file is the canonical source of how to build Network Service Mesh in case this file becomes out of date.
+
```
Comment (denis-tingajkin) [inline docs/guide-build.md]: I've updated the `guide-build.md`, please take a look.

File: .circleci/config.yml [hunk 5]
```diff
@@ -51,7 +46,22 @@ jobs:
     environment:
       GO111MODULE: "on"
 
-# packet
+  # code analysis
+  static-code-analysis:
```
Comment (denis-tingajkin) [inline .circleci/config.yml]: @nickolaev I've added job with `static-code-analysis` and got strange result: https://circleci.com/gh/networkservicemesh/networkservicemesh/46875?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link
```
Received 'killed' signal
```
Are you know why this happens?

File: .circleci/config.yml [hunk 6]
```diff
@@ -22,11 +22,6 @@ jobs:
...
             ./scripts/go-mod-download.sh
       - run: "! go fmt ./... 2>&1 | read"
-      - run:
-          name: golint
-          command: |
-            go get golang.org/x/lint/golint
-            golint ./...
 #      - run:
 #          name: gosec
 #          command: |
```
Comment (nickolaev) [inline .circleci/config.yml]: gosec can go away too, please remove it.

File: .circleci/config.yml [hunk 7]
```diff
@@ -51,7 +46,22 @@ jobs:
...
+  static-code-analysis:
+    resource_class: medium+
+    docker:
+      - image: circleci/golang
+    steps:
+      - checkout
+      - run:
+          name: lint
+          command: |
+            go get github.com/golangci/golangci-lint/cmd/golangci-lint
```
Comment (nickolaev) [inline .circleci/config.yml]: Please use a released binary instead:
https://github.com/golangci/golangci-lint/releases/download/v1.16.0/golangci-lint-1.16.0-linux-amd64.tar.gz

We will avoid ambiguous results and have reproducible error failures.

File: .circleci/config.yml [hunk 8]
```diff
@@ -549,7 +559,10 @@ workflows:
   build-and-test:
     jobs:
       - sanity-check
-
+      - static-code-analysis:
+          name: "static-code-analysis"
+          requires:
+            - sanity-check
```
Comment (nickolaev) [inline .circleci/config.yml]: OK!

File: docs/guide-build.md [hunk 9]
```diff
@@ -206,3 +206,24 @@ If you want to run it locally, you need to [install shellcheck](https://github.c
...
+Checking changes:
+```bash
+golangci-lint run --new-from-rev=origin/master
+```
+Checking all code in the project:
+```bash 
+golangci-lint run ./...
+```
+If you have any unsolvable problem with a concrete linter then consider updating `.golanci.yaml`
```
Comment (nickolaev) [inline docs/guide-build.md]: Good!

File: .circleci/config.yml (hunk 6 above)
Comment (denis-tingajkin) [reply .circleci/config.yml]: thanks, fixed

File: .circleci/config.yml (hunk 7 above)
Comment (denis-tingajkin) [reply .circleci/config.yml]: thanks, fixed

File: docs/guide-build.md (hunk 9 above)
Comment (haiodo) [reply docs/guide-build.md]: Could we have make lint-install, make lint-check and make lint-fix?

Comment (denis-tingajkin) [reply docs/guide-build.md]: @haiodo  good idea, fixed!

**Comment by edwarnicke** [comment]
@denis-tingajkin Why so much import reordering?  Most files touched just appear to have their import orders and whitespace shifted.

**Comment by denis-tingajkin** [comment]
@edwarnicke It's auto fixing by `goimports` and `misspelling` linters. You can also do a fix with ```make lint-fix``` after this PR merged.
 I'll provide another PR with the most critical fixes with https://github.com/networkservicemesh/networkservicemesh/issues/1105

**Pull request merged**